### ISSUE TITLE
armv7 (armv8/aarch32?): deadlock spinning in optee core

### ISSUE THREAD

**Issue by etienne-lms** [OP-TEE/optee_os#1099]
Dear all,

I faced a deadlock issue running optee on a armv7/SMP SoC.
To reproduce such an issue, one must have all cores trying to lock some optee spinlock at the same time. I found the best chances to reproduce it was to run only 2 cores in SMP and endlessly execute the "`xtest 1013`".

Once issue occurs, i found that one core requested a `cpu_spin_lock()` over an locked spinlock. Hence it waits for an event before spinning again. The other core is calling  `cpu_spin_lock()` over an unlocked spinlock: it executes a `strex.` For some reason i could not find yet, the `strex` failed (it did not get exclusive access). Hence the other waits for an even before spinning again.

As both available cores are each waiting for event with IRQs masked... both get stucked.

Does one think this is a relevant bug ? Maybe i have made some mis-configuration elsewhere (I could not found yet).

I could not reproduce the deadlock on a aarch64 optee. Looking at the aarch64 `cpu_spin_lock()` code, failure to get the exclusive access does not trig a 'wait-for-event' before spinning back . I did not try armv8/aarch32, but i think we should end in the same deadlock scheme.

As appendix, i think the patch should fix the issue but i would rather have some smart guys confirming there really is an issue. One line to remove in arch/arch/kernel/proc_a32.S:
```
 cpu_spin_lock:
     mov r2, #SPINLOCK_LOCK
 _spinlock_loop:
     ldrex r1, [r0]
     cmp r1, #SPINLOCK_UNLOCK
     wfene
     bne _spinlock_loop
     strex r1, r2, [r0]
     cmp r1, #0
-    wfene
     bne _spinlock_loop
     dmb
     bx lr

```

**Comment by jenswi-linaro** [comment]
Please compare with the Aarch32 implementation in ARM-TF https://github.com/ARM-software/arm-trusted-firmware/blob/master/lib/locks/exclusive/aarch32/spinlock.S#L37

As we've based our Aarch64 implementation on the corresponding implementation in ARM-TF I think we should do the same with our Aarch32 implementation now that one is available in ARM-TF.

**Comment by etienne-lms** [comment]
ok, i will sync with a-tf and push a PR for that. 
i see it does not 'wfe' on 'strex' failure.

**Comment by etienne-lms** [comment]
see #1101

**Comment by etienne-lms** [comment]
fix #1101 merged.
